xls/modules/zstd/zstd_dec_test.cc

```diff
@@ -0,0 +1,297 @@
+// Copyright 2024 The XLS Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <algorithm>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <filesystem>  // NOLINT
+#include <iomanip>
+#include <ios>
+#include <iostream>
+#include <iterator>
+#include <memory>
+#include <optional>
+#include <ostream>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/status/statusor.h"
+#include "absl/types/span.h"
+#include "external/com_github_facebook_zstd/lib/zstd.h"
```

Comment:
I don't think we have other example of referencing to an header using the raw bazel workspace in the path, is there a better way to export this here?

Reply:
we have two examples, I mentioned above. Thisis the recommended way currently in our repo.

Reply:
The name `com_github_facebook_zstd` might be a bit long. Note, we also already have 
[imported zstd under the name `net_zstd`](https://github.com/google/xls/blob/9941c71fa6c3a74cf45a7bbbe3c0991da32e9ee0/dependency_support/load_external.bzl#L196-L202) which is a recent version.

To avoid weird diamond dependencies, maybe we should just use that one ? Or is it crucial for this change to use the older version [from 2020](https://github.com/facebook/zstd/releases/tag/v1.4.7) ?

Reply:
+1 for updating to 1.5.6, as it matches the internal version.

Reply:
There are examples provided by @hzeller [in this comment](https://github.com/google/xls/pull/1315#discussion_r1679660926). It so happens that those examples have simpler dependency names (without `com_github`, etc.).
The prefix `com_github_facebook_zstd` comes from the name of the `http_archive` bazel rule to import the zstd dependency in `dependency_support/load_external.bzl`. We can change that to `zstd` so that in BUILD files we would do:
```
- "@com_github_facebook_zstd//:zstd",
+ "@zstd//:zstd",
```

And in C++ files:
```
- #include "external/com_github_facebook_zstd/lib/zstd.h"
+ #include "external/zstd/lib/zstd.h"
```

@proppy please let me know if that is OK for you.

Reply:
I think @hzeller was suggesting to depend on the existing https://github.com/google/xls/blob/9941c71fa6c3a74cf45a7bbbe3c0991da32e9ee0/dependency_support/load_external.bzl#L196-L202 dep we have rather than introducing a new one.

Reply:
Right, I will have to make sure that the BUILD file used there exposes `decodecorpus` and correct headers. Most probably I will also have to modify the patch for decodecorpus to match with libzstd 1.5.6v codebase

Reply:
`net_zstd` is not referenced anywhere in the XLS workspace. I'm aware that this is:
* a transitive dependency of [riegeli](https://github.com/google/riegeli/blob/411cda7f6aa81f8b8591b04cf141b1decdcc928c/WORKSPACE#L21-L27) in [fuzztest](https://github.com/google/fuzztest/blob/393ae75c0fca5f9892e73969da5d6bce453ad318/WORKSPACE#L86-L92) - this version is still WORKSPACE-based and uses old version of `riegeli` which was also WORKSPACE-based so it was required to define `net_zstd` there as we need to import all transitive deps in WORKSPACE-based bazel workspaces
* a dependency of [riegeli](https://github.com/google/riegeli/blob/cb68d579f108c96831b6a7815da43ff24b4e5242/MODULE.bazel#L66-L70)  in [XLS](https://github.com/google/xls/blob/88c10c18b7b6db811796c267ffccd0499210e226/dependency_support/load_external.bzl#L188-L193) - `riegeli` version used in XLS already uses bzlmod and there is no need for defining transitive dependencies like `net_zstd`.

To further confirm that `net_zstd` defined in XLS is not used I defined and explicit dependency on it in one of ZSTD decoder tests. It turned out that the path to `build_file`  (`build_file = "@com_google_riegeli//third_party:net_zstd.BUILD"` ) is invalid, see the error:
```
ERROR: An error occurred during the fetch of repository 'net_zstd':                                                                                                                                                                                                                         
   Traceback (most recent call last):                                                                                                                                                                                                                                                       
        File "/data/.cache/bazel/_bazel_pczarnecki/9fcaa4a65d2ba702fc1ad78cc6dec096/external/bazel_tools/tools/build_defs/repo/http.bzl", line 141, column 28, in _http_archive_impl                                                                                                        
                workspace_and_buildfile(ctx)                                                                                                  
        File "/data/.cache/bazel/_bazel_pczarnecki/9fcaa4a65d2ba702fc1ad78cc6dec096/external/bazel_tools/tools/build_defs/repo/utils.bzl", line 68, column 41, in workspace_and_buildfile                                                                                                   
                ctx.file("BUILD.bazel", ctx.read(ctx.attr.build_file))                                                                        
Error in read: Unable to load package for @@com_google_riegeli//third_party:net_zstd.BUILD: BUILD file not found in directory 'third_party' of external repository @@com_google_riegeli. Add a BUILD file to a directory to mark it as a package.                                           
ERROR: no such package '@@net_zstd//': Unable to load package for @@com_google_riegeli//third_party:net_zstd.BUILD: BUILD file not found in directory 'third_party' of external repository @@com_google_riegeli. Add a BUILD file to a directory to mark it as a package.                   
ERROR: /data/work/google-xls/xls/xls/modules/zstd/BUILD:201:8: //xls/modules/zstd:frame_header_cc_test depends on @@net_zstd//:net_zstd in repository @@net_zstd which failed to fetch. no such package '@@net_zstd//': Unable to load package for @@com_google_riegeli//third_party:net_zst
d.BUILD: BUILD file not found in directory 'third_party' of external repository @@com_google_riegeli. Add a BUILD file to a directory to mark it as a package.                                                                                                                              
ERROR: Analysis of target '//xls/modules/zstd:frame_header_cc_test' failed; build aborted: Analysis failed                                    
INFO: Elapsed time: 5.011s, Critical Path: 0.02s                                                                                                                                                                                                                                            
INFO: 1 process: 1 internal.                                                                                                                                                                                                                                                                
ERROR: Build did NOT complete successfully                                                                                                                                                                                                                                                  
ERROR: Build failed. Not running target
```

The path to build file is a path that refers to older versions of `riegeli` (the one used in `fuzztest` workspace). In XLS workspace bazel will refer to `@com_google_riegeli` defined in XLS workspace which has a never version and it does not contain build file for ZSTD at all.

I'm not sure why this `net_zstd` dependency was introduced to the XLS workspace as it is not used anywhere because it would fail during the build as you can see above.

@proppy @hzeller 
What do you think about removing the `@net_zstd` dependency, keeping the `@com_github_facebook_zstd`, renaming it to shorter `@zstd` and bumping the version to `1.5.6`?
We can copy the [BUILD.bazel](https://github.com/bazelbuild/bazel-central-registry/blob/main/modules/zstd/1.5.6/patches/add_build_file.patch) from the upstream bazel central registry and apply a patch that exposes `decodecorpus` utility and modifies it to enable ZSTD decoder tests.

Reply:
> What do you think about removing the @net_zstd dependency, keeping the @com_github_facebook_zstd, renaming it to shorter @zstd and bumping the version to 1.5.6?

That seems like a good idea! Ideally this will all get easier once/if we switch to bzlmod.

Reply:
I think using the `@zstd` name sounds good (which is also the name used in the repo once we switch to bzlmod https://registry.bazel.build/modules/zstd ).

The net_zstd thing sais that it is used by fuzztest. Are you saying we can remove it and fuzztest still works ? If so: yes, lets do it. If not, we still can add a `repo_mapping` to point to the `@zstd`.

Reply:
It looks like we can safely remove `@net_zstd`. I did that, reintroduced `@com_github_facebok_zstd`, renamed it to `@zstd`, and bumped version from `v1.4.7` to `v1.5.6`. The build file was imported from the [bazel central registry](https://github.com/bazelbuild/bazel-central-registry/blob/3656ebca9df070b71bf90a88ac16fc1e7b06b585/modules/zstd/1.5.6/patches/add_build_file.patch).